modules/airgridRtdProvider.js

```diff
@@ -80,28 +81,42 @@ function setAudiencesToAppNexusAdUnits(adUnits, audiences) {
 
 /**
  * Pass audience data to configured bidders, using ORTB2
+ * @param {Object} bidConfig
  * @param {Object} rtdConfig
  * @param {Array} audiences
- * @return {{}} a map from bidder code to ORTB2 config
+ * @return {void}
  */
-export function setAudiencesAsBidderOrtb2(rtdConfig, audiences) {
+export function setAudiencesAsBidderOrtb2(bidConfig, rtdConfig, audiences) {
   const bidders = deepAccess(rtdConfig, 'params.bidders');
   if (!bidders || bidders.length === 0 || !audiences || audiences.length === 0) return;
 
-  const keywords = audiences.map(
+  const agOrtb2 = {};
+
+  const agKeywords = audiences.map(
     (audienceId) => `perid=${audienceId}`
   ).join(',');
+  deepSetValue(agOrtb2, 'ortb2.user.keywords', agKeywords);
 
-  config.mergeBidderConfig({
-    bidders: bidders,
-    config: {
-      ortb2: {
-        site: {
-          keywords,
+  // Is this correct?
+  const agUserData = [
+    {
+      id: AG_TCF_ID, // Not sure of this value
+      name: 'airgrid', // Not sure of this value
+      segment: [
+        {
+          id: 'perid', // Not sure of this value
+          name: 'perid', // Not sure of this value
+          value: audiences,
         }
-      }
+      ]
     }
-  })
+  ]
+  deepSetValue(agOrtb2, 'ortb2.user.data', agUserData);
+
+  const bidderConfig = Object.fromEntries(
+    bidders.map((bidder) => [bidder, agOrtb2])
+  )
+  mergeDeep(bidConfig?.ortb2Fragments?.bidder, bidderConfig)
```

Comment:
Not really sure of the spec here, so I'm doing some guess-work.
As I understand, this includes exchange specific values, so we need to check with xandr guys if we're getting this right?
Do you happen to know anything about those values, @ydennisy?

Also, @dgirardi, do you think it is ok for us to just set the values in the `agOrtb2` obj and merge it in the `bidConfig.ortb2Fragments.bidder` like so?
I saw it [in some other module](https://github.com/prebid/Prebid.js/blob/18f9774c7a9d989594ef6e6ef78231d0c80c0b3f/modules/permutiveRtdProvider.js#L165) where they actually access previous data and merge it in the array prior to setting, but they are doing some different stuff there and I'm not sure it applies here.

Reply:
this looks OK, except that there's an extra `ortb2` - with your setup I believe line 98 should be `  deepSetValue(agOrtb2, 'user.keywords', agKeywords);`

You can test this by looking at what gets into bid requests; if you configure your module and run an auction, `pbjs.onEvent('bidRequested', (br) => console.log(br.ortb2))`  should show your data for the bidders that are set up to get it.

Reply:
@dgirardi, thanks for the directions!

Yesterday, I tried and tested both using `ortb2.user.keywords` and `user.keywords` for setting the ortb2 data, but I was unable to check in our [integration example page](https://github.com/prebid/Prebid.js/blob/master/integrationExamples/gpt/airgridRtdProvider_example.html) that module was behaving correctly.
The data I got back from `pbjs.onEvent('bidRequested', (br) => console.log(br.ortb2))` was the same as the one set in the `pbjs.setBidderConfig` method in that page, and I couldn't check that it was being mutated with the configs we pass in the `bidConfig?.ortb2Fragments?.bidder`.

Maybe I'm not getting something fully?
Do you happen to have any good example that I could use as reference on how to properly test the integration, specially in an automated manner?
We'd like to be really sure that the module is not broken as it is critical for us.

Also, sorry for being so asking, but maybe you do know about the validity of the data being passed to `user.data` with the segments there or could point me to some docs about it?
I also looked around and could only find [one example](https://github.com/prebid/Prebid.js/blob/4d9b28c8bd7a19589d4d4e9b4b5925d35e52ac00/integrationExamples/gpt/permutiveRtdProvider_example.html#L202-L207) of another module setting that data, it is not very complete, and when trying and looking with `pbjs.onEvent`, their data actually returns always empty.

Reply:
your integration example does not set up `params.bidders` in the data provider config so it shorts out at line 91 in this file. With it, and the `ortb2` change, it seems to work:

![image](https://github.com/AirGrid/Prebid.js/assets/10151381/39aba173-6d71-4e9d-bfc5-c5bc2e4771e0)

regarding the format, it looks right to me in  the sense that it conforms to spec, but I know very little about how the taxonomies are managed. The wider group @ Prebid would be a better target for the question, probably easier to get people involved from the Prebid.js PR board. I'll say that if in practice your only target is appnexus, I think what you have here is more than enough (with the fixes mentioned above).

Reply:
actually, I didn't pay enough attention - some `data[].id` and `data[].id.segment[].value` should be strings, so I'd encode the example above as:

```json
{
  "id": "782",
  "name": "airgrid",
  "segment": [
     {
       "id": "perid",
       "name": "perid",
       "value": "sport",
     },
     {
       "id": "perid",
       "name": "perid",
       "value": "travel",
     }
  ]
}
```

Reply:
Hey @dgirardi, thank you very much!
I got it to work adding the `params.bidders`. I've also fixed the `segment` to conform with the spec. Thank you for the catch!